Add support for Windows instances

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR adds support for Windows instances to the configuration for assessment environments. This includes:
- Adding a Windows instance configuration
- Adding any supporting pieces such as an instance profile, A record, ACL rules, etc
- Rename the `vnc_parameterstorereadonly` policy/role resources to reflect that they are not VNC-only 
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

Assessors need access to Windows instances as part of their operations.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
Automated tests pass.

I tested spinning up a Windows instance, verified I could access it through Guacamole, and verified that with a Samba instance in the environment the share was connectable.
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

### Known issues ###

The mounted Samba share shows as `Disconnected`, but works fine when access is attempted. I believe this is an issue with how the share is mounted and the permissions of the connected user.

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

mcdonnnj committed:
Add ingress via RDP to Guacamole Accessible SG

This allows both TCP and UDP traffic for RDP.

mcdonnnj committed:
Update private ACL rules to allow RDP egress

This allows RDP (both TCP and UDP) egress to the operations subnet from the
private subnet CIDRs.

mcdonnnj committed:
Update operations ACL rules to allow RDP ingress

This allows RDP (both TCP and UDP) ingress to the operations subnet from the
private subnet CIDRs.

mcdonnnj committed:
Add configuration to support Windows instances

This supports spinning up Windows instances in assessment environments and
includes the necessary IAM, Route53, and Security Group configurations.

mcdonnnj committed:
Add support for Windows to Guacamole cloud-init

Add support for Windows instances to the Guacamole cloud-init configuration.
This included adjusting the SQL template, the script to render the template,
and the appropriate variables and documentation to support these changes.

mcdonnnj committed:
Read in the connection SQL template once

Adjust the script that renders SQL templates for Guacamole connections to only
read in the SQL template once instead of for every connection. This data is
used as-is with no changes made to it so it doesn't make sense to read in the
template data for every connection. Instead we read it in once and use the same
template data for each connection's generated file.

mcdonnnj committed:
Add Windows instance resources to outputs

Add certain resources related to Windows instance to the outputs for the
project. The resources match the resources for other instance types that
are also outputs (instance profile, instances, security group).

mcdonnnj committed:
Adjust Windows instance Security Groups

Add the smb_client security group to allow access to SMB shares on Samba
instances in the environment. Fix the order of security groups so they are
correctly alphabetic.

mcdonnnj committed:
Add read permissions for RDP ParameterStore keys

Add `/rdp/*` to the resources that the ParameterStoreReadOnly role is allowed
to read. This will allow access to the values necessary to configure Guacamole
configuration.

mcdonnnj committed:
Update references for Guacamole Parameter Store access

Update policy and role references to replace VNC references with Guacamole.
This is to better reflect the fact that VNC is not the only connection
method in use with the addition of RDP for Windows instances.

mcdonnnj committed:
Update assume role policy to reflect naming changes

Update the assume role to mirror the naming changes made to the role and policy
used to access Guacamole related Parameter Store keys.

mcdonnnj committed:
Update README to reflect resource name changes

mcdonnnj committed:
Add user_data for Windows instances

Add an EC2Launch v2 configuration template and use it to populate the user_data
for Windows instances.

mcdonnnj committed:
DRY out two security group rules

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

mcdonnnj committed:
Merge branch 'develop' into improvement/add_windows_instances

mcdonnnj committed:
Update Windows instance block device settings

Update to include the changes made to other instances in
cisagov/cool-assessment-terraform#148.

jsf9k commented on nessus_sg.tf:
```diff
@@ -37,6 +37,20 @@ resource "aws_security_group_rule" "nessus_ingress_from_kali_via_web_ui" {
   to_port                  = 8834
 }
 
+# Allow ingress from Windows instances via Nessus web GUI
+#
+# For: Assessment team Nessus web access from Windows instances
+resource "aws_security_group_rule" "nessus_ingress_from_windows_via_web_ui" {
+  provider = aws.provisionassessment
+
+  security_group_id        = aws_security_group.nessus.id
+  type                     = "ingress"
+  protocol                 = "tcp"
+  source_security_group_id = aws_security_group.windows.id
+  from_port                = 8834
+  to_port                  = 8834
+}
+
```
I don't think we want this security group rule in place.  The Windows instances probably don't need to access Nessus.

dav3r replied:
I agree.  If someone asks for this functionality, we can always add it in later, but I don't think anyone intends to hit Nessus from the Windows instances.

jsf9k replied:
@justinforbes states that he does indeed want this access.

jsf9k commented on pentestportal_sg.tf:
```diff
@@ -25,6 +25,22 @@ resource "aws_security_group_rule" "pentestportal_ingress_from_kali_via_web" {
   to_port                  = each.key
 }
 
+# Allow ingress from Windows instances via ports 443 and 8080
+#
+# For: Assessment team HTTPS access to PenTest Portal (development and
+# production modes) from Windows instances
+resource "aws_security_group_rule" "pentestportal_ingress_from_windows_via_web" {
+  for_each = toset(["443", "8080"])
+  provider = aws.provisionassessment
+
+  security_group_id        = aws_security_group.pentestportal.id
+  type                     = "ingress"
+  protocol                 = "tcp"
+  source_security_group_id = aws_security_group.windows.id
+  from_port                = each.key
+  to_port                  = each.key
+}
+
```
I don't think we want this security group rule.  The Windows instances should not require access to Pentest Portal instances.

dav3r replied:
Agreed.

jsf9k replied:
@justinforbes states that he does indeed want this access.

jsf9k commented on variables.tf:
```diff
@@ -174,3 +174,9 @@ variable "terraformer_role_name" {
+
+variable "windows_with_docker" {
+  type        = bool
+  description = "A boolean to control the instance type used when creating Windows instances to allow Docker Desktop support."
```
There should be a second sentence to the description stating that bare metal instance types are required to allow Docker Desktop support.  Without that key piece of information it's unclear to most folks what Docker Desktop has to do with the instance type.

mcdonnnj replied:
Please see 3d97ef5 and let me know if it is sufficient.

jsf9k commented on windows_sg.tf:
```diff
@@ -0,0 +1,52 @@
+  }
+}
+
+# Allow egress to PenTest Portal instances via ports 443 and 8080
+resource "aws_security_group_rule" "windows_egress_to_pentestportal_via_web" {
+  for_each = toset(["443", "8080"])
+  provider = aws.provisionassessment
+
+  security_group_id        = aws_security_group.windows.id
+  type                     = "egress"
+  protocol                 = "tcp"
+  source_security_group_id = aws_security_group.pentestportal.id
+  from_port                = each.key
+  to_port                  = each.key
+}
+
+# Allow egress to Nessus instances via port 8834 (the default port
+# used by the Nessus web UI)
+resource "aws_security_group_rule" "windows_egress_to_nessus_via_web_ui" {
+  provider = aws.provisionassessment
+
+  security_group_id        = aws_security_group.windows.id
+  type                     = "egress"
+  protocol                 = "tcp"
+  source_security_group_id = aws_security_group.nessus.id
+  from_port                = 8834
+  to_port                  = 8834
+}
```
We don't want these security group rules in place, for the reasons explained above.

jsf9k replied:
@justinforbes states that he does indeed want this access.

jsf9k commented on windows_ec2.tf:
```diff
@@ -0,0 +1,69 @@
+  vpc_security_group_ids = [
+    aws_security_group.cloudwatch_and_ssm_agent.id,
+    aws_security_group.guacamole_accessible.id,
+    aws_security_group.scanner.id,
```
I don't know if this kind of ingress/egress is required.  We should ask the assessors.

If they only need HTTP and HTTPS egress, for example, then I'd prefer to add those rules to the Windows security group and drop the scanner security group.

dav3r replied:
I'd lean towards giving them the minimum and then add more access if requested.

jsf9k replied:
@justinforbes states that he does indeed want this access.

jsf9k reviewed (changes requested):
We need to confirm a few things with the assessors, but I'm pretty sure some of the allowed ingress and egress can be gotten rid of.

mcdonnnj committed:
Improve variable description

Improve the description of the windows_with_docker variable to better explain
what it controls.

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

dav3r commented on ec2launch/windows-setup.tpl.yml:
```diff
@@ -0,0 +1,21 @@
+          $samba_server = ("${samba_server_input}" -split ",")[0]
+          if (
+            ($samba_server -ne "") -and
+            (-not (Get-PSDrive "N" -ErrorAction "SilentlyContinue"))
+          )
+          {
+            New-PSDrive -Name "N" `
```
Am I correct in understanding that `"N"` is the drive letter for the Samba instance here?  If so, should this be an input variable to this script, rather than a hardcoded, somewhat magical value?

mcdonnnj replied:
I can change it to an input. I went with a hard-coded value only because I didn't expect there to be much change.

mcdonnnj replied:
You may enjoy 1ff9ebb.

dav3r replied:
Lovely, thanks!

dav3r commented on locals.tf:
```diff
@@ -199,9 +199,9 @@ locals {
   # access to the VNC-related SSM Parameter Store parameters in the
   # Images account.
-  vnc_parameterstorereadonly_role_description = format("Allows read-only access to VNC-related SSM Parameter Store parameters required for the %s assessment.", var.assessment_account_name)
+  gucamole_parameterstorereadonly_role_description = format("Allows read-only access to VNC-related SSM Parameter Store parameters required for the %s assessment.", var.assessment_account_name)
```
```suggestion
  gucamole_parameterstorereadonly_role_description = format("Allows read-only access to VNC-related and RDP-related SSM Parameter Store parameters required for the %s assessment.", var.assessment_account_name)
```

dav3r commented on locals.tf:
```diff
@@ -199,9 +199,9 @@ locals {
+  gucamole_parameterstorereadonly_role_description = format("Allows read-only access to VNC-related SSM Parameter Store parameters required for the %s assessment.", var.assessment_account_name)
 
-  vnc_parameterstorereadonly_role_name = format("ParameterStoreReadOnly-%s-VNC", local.assessment_workspace_name)
+  gucamole_parameterstorereadonly_role_name = format("ParameterStoreReadOnly-%s-VNC", local.assessment_workspace_name)
```
```suggestion
  gucamole_parameterstorereadonly_role_name = format("ParameterStoreReadOnly-%s-VNC-RDP", local.assessment_workspace_name)
```

dav3r commented on locals.tf:
```diff
@@ -199,9 +199,9 @@ locals {
   # The name and description of the role and policy that allows read-only
   # access to the VNC-related SSM Parameter Store parameters in the
   # Images account.
```
Update comment to reflect addition of RDP:
```suggestion
  # access to the VNC-related and RDP-related SSM Parameter Store
  # parameters in the Images account.
```

This change should be batched with similar changes below.

dav3r commented on windows_route53.tf:
```diff
@@ -0,0 +1,11 @@
+# Private DNS A record for Windows instance
```
Minor, but slightly more accurate:
```suggestion
# Private DNS A record for Windows instances
```

mcdonnnj replied:
Oops, I copied that from `kali_route53.tf` and didn't notice.

dav3r reviewed (changes requested):
Great job here @mcdonnnj!  Most of my suggestions are minimal, but please take a look at them.  I think @jsf9k covered the biggest thing, which is disallowing connectivity between Windows and the Nessus/PenTestPortal instances.

mcdonnnj committed:
Update descriptions for Guacamole SSM access

Update comments, descriptions, and a role name to reflect that access is for
both VNC and RDP-related SSM paramters.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj committed:
Fix comment about private DNS records

The resource manages DNS A records for multiple instances, not a single instance.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj committed:
Make PSDrive drive letter a template variable

Instead of hard-coding the drive letter for the share's PSDrive we should make
it a variable for the template. This will allow customization if needed while
not impacting the current implementation with the default value.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k reviewed (approved):
Potent work!

mcdonnnj committed:
Add rules to allow access between instance types

Allow unfettered access between Kali and Windows instances. This allows both
TCP and UDP traffic between them over any port.

dav3r reviewed (approved):
Looks excellent! 🦅 🚀

jsf9k reviewed (approved):
_Approval intensifies!!!_
